File: MathsThing_Plugins/src/uk/ac/uea/mathsthing/plugins/functions/ATan.java
package uk.ac.uea.mathsthing.plugins.functions;

import java.math.BigDecimal;

import uk.ac.uea.mathsthing.IFunctionPlugin;
import uk.ac.uea.mathsthing.util.FormulaException;

public class ATan extends IFunctionPlugin {
	
	@Override
	public String getName()
	{
		return "atan";
	}

	@Override
	public BigDecimal function(BigDecimal input) throws FormulaException
	{
		BigDecimal bd = new BigDecimal(Math.atan(input.doubleValue()));
		// This is to stop abnormally large results being returned as a result of the asymptote.
		if (bd.doubleValue() > 100 || bd.doubleValue() < -100) return null;
		else return bd;
	}

	
}

File: MathsThing_Plugins/src/uk/ac/uea/mathsthing/plugins/functions/ATan.scala
package uk.ac.uea.mathsthing.plugins.functions

import uk.ac.uea.mathsthing.IPlugin.IFunctionPlugin
import java.math.BigDecimal

class ATan extends IFunctionPlugin {

  def getName : String = "atan"
    
  def function(input: BigDecimal) : BigDecimal = {
  	if (input.doubleValue == Double.NaN 
  	    || input.doubleValue <= Double.NegativeInfinity 
  	    || input.doubleValue >= Double.PositiveInfinity)
      null
    
    try {
      val result = new BigDecimal(Math.atan(input.doubleValue))
      
      if (result.doubleValue == Double.NaN 
  	      || result.doubleValue <= Double.NegativeInfinity 
  	      || result.doubleValue >= Double.PositiveInfinity 
  	      || result.doubleValue > 1000 || result.doubleValue < -1000)
        null;
  	  else
        result
    } catch {
      case ex:NumberFormatException => null
    }
  }
}